### PR TITLE
[WebAssembly] Support the new "Lime1" CPU

### PR DESCRIPTION
**Pull request by sunfishcode**
This adds WebAssembly support for the new [Lime1 CPU].

First, this defines some new target features. These are subsets of existing
features that reflect implementation concerns:

 - "call-indirect-overlong" - implied by "reference-types"; just the overlong
   encoding for the `call_indirect` immediate, and not the actual reference
   types.

 - "bulk-memory-opt" - implied by "bulk-memory": just `memory.copy` and
   `memory.fill`, and not the other instructions in the bulk-memory
    proposal.

Next, this defines a new target CPU, "lime1", which enables mutable-globals,
bulk-memory-opt, multivalue, sign-ext, nontrapping-fptoint, extended-const,
and call-indirect-overlong. Unlike the default "generic" CPU, "lime1" is meant
to be frozen, and followed up by "lime2" and so on when new features are
desired.

[Lime1 CPU]: https://github.com/WebAssembly/tool-conventions/blob/main/Lime.md#lime1

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -854,9 +854,15 @@ and `-mbulk-memory` flags, which correspond to the [Bulk Memory Operations]
 and [Non-trapping float-to-int Conversions] language features, which are
 [widely implemented in engines].
 
+A new Lime1 target CPU is added, -mcpu=lime1. This CPU follows the definition of
+the Lime1 CPU [here], and enables -mmultivalue, -mmutable-globals,
+-mcall-indirect-overlong, -msign-ext, -mbulk-memory-opt, -mnontrapping-fptoint,
```
Comment (sbc100) [inline clang/docs/ReleaseNotes.rst:859]: What is the rationale for including the overlong call indirect here? 

Would it be unreasonable to suggest that we land the two new features in a separate PR before landing the new CPU type that uses them?  (Maybe its hard to split up and test?)

Comment (sunfishcode) [reply clang/docs/ReleaseNotes.rst:859]: I've now split out those parts into https://github.com/llvm/llvm-project/pull/117087.

**Review by sbc100 — APPROVED** [review]

**Comment by dschuff** [comment]
LGTM too; I just noticed that nowhere does the tool-conventions document mention that "lime" is short for "linear memory" and that the intention is that this is the subset of features useful to toolchains or runtimes that focus on linear memory. Maybe we should add that.

**Comment by sunfishcode** [comment]
That feels to me like something best documented in the definition of Lime itself, rather than in LLVM. I've now opened https://github.com/WebAssembly/tool-conventions/pull/241 to document the inspiration for the name.

File: clang/lib/Basic/Targets/WebAssembly.cpp [hunk 2]
```diff
@@ -167,6 +167,17 @@ bool WebAssemblyTargetInfo::initFeatureMap(
...
+  auto addLime1Features = [&]() {
+    // Lime1:
+    // <https://github.com/WebAssembly/tool-conventions/blob/main/Lime.md#lime1>
+    Features["multivalue"] = true;
+    Features["mutable-globals"] = true;
+    Features["call-indirect-overlong"] = true;
+    Features["sign-ext"] = true;
+    Features["bulk-memory-opt"] = true;
+    Features["nontrapping-fptoint"] = true;
+    Features["extended-const"] = true;
```
Comment (aheejin) [inline clang/lib/Basic/Targets/WebAssembly.cpp:179]:
```suggestion
    Features["bulk-memory-opt"] = true;
    Features["call-indirect-overlong"] = true;
    Features["extended-const"] = true;
    Features["multivalue"] = true;
    Features["mutable-globals"] = true;
    Features["nontrapping-fptoint"] = true;
    Features["sign-ext"] = true;
```
Sorry alphabetizing police again...

File: llvm/lib/Target/WebAssembly/WebAssembly.td [hunk 3]
```diff
@@ -127,6 +127,13 @@ def : ProcessorModel<"generic", NoSchedModel,
                        FeatureMutableGlobals, FeatureNontrappingFPToInt,
                        FeatureReferenceTypes, FeatureSignExt]>;
 
+// Lime1: <https://github.com/WebAssembly/tool-conventions/blob/main/Lime.md#lime1>
+def : ProcessorModel<"lime1", NoSchedModel,
+                      [FeatureMultivalue, FeatureMutableGlobals,
+                       FeatureCallIndirectOverlong, FeatureSignExt,
+                       FeatureBulkMemoryOpt, FeatureNontrappingFPToInt,
+                       FeatureExtendedConst]>;
```
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssembly.td:135]:
```suggestion
                      [FeatureBulkMemoryOpt, FeatureCallIndirectOverlong,
                       FeatureExtendedConst, FeatureMultivalue,
                       FeatureMutableGlobals, FeatureNontrappingFPToInt,
                       FeatureSignExt]>;
```

**Review by aheejin — APPROVED** [review]
